File: java/105_ConstructBinaryTreeFromPreorderAndInorderTraversal.java
class Solution {
    int p, i;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        p = 0;
        i = 0;
        return build(Integer.MIN_VALUE, preorder, inorder);
    }
    
    public TreeNode build(int stop, int[] preorder, int[] inorder) {
        if (p >= preorder.length || inorder[i] == stop) 
            return null;
        
        TreeNode root = new TreeNode(preorder[p]);
        p++;
        
        root.left = build(root.val, preorder, inorder);
        i++;

        root.right = build(stop, preorder, inorder);
        return root;
    }
}

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorderArray, int[] inorderArray) {
        Deque<Integer> preorder = new ArrayDeque<Integer>();
        Deque<Integer> inorder = new ArrayDeque<Integer>();
        for (int i = 0; i < inorderArray.length; ++i) {
            preorder.addFirst(preorderArray[i]);
            inorder.addFirst(inorderArray[i]);
        }
        return build(preorder, inorder, Integer.MIN_VALUE);
    }
    
    public TreeNode build(Deque<Integer> preorder, Deque<Integer> inorder, int stop) {
        if (preorder.isEmpty() || inorder.getLast() == stop)
            return null;
        TreeNode root = new TreeNode(preorder.removeLast());
        root.left = build(preorder, inorder, root.val);
        inorder.removeLast();
        root.right = build(preorder, inorder, stop);
        return root;
    }
}

File: cpp/105_ConstructBinaryTreeFromPreorderAndInorderTraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        reverse(preorder.begin(), preorder.end());
        reverse(inorder.begin(), inorder.end());

        return build(preorder, inorder, INT_MIN);
    }
    
    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int stop) {
        if (preorder.empty() || inorder[inorder.size()-1] == stop)
            return NULL;
        
        TreeNode* root = new TreeNode(preorder.back());
        preorder.pop_back();
        
        root->left = build(preorder, inorder, root->val);
        inorder.pop_back();
        root->right = build(preorder, inorder, stop);
        return root;
    }
};
